api/src/importers/cmdb/datasource.js

```diff
@@ -0,0 +1,162 @@
+const defaultLog = require('../../utils/logger')('agri-csv-datasource');
+const RECORD_TYPE = require('../../utils/constants/record-type-enum');
+
+class CmdbCsvDataSource {
+  /**
+   * Creates an instance of DataSource.
+   *
+   * @param {*} taskAuditRecord audit record hook for this import instance
+   * @param {*} auth_payload information about the user account that started this update
+   * @param {*} recordType record type to create from the csv file
+   * @param {*} csvRows array of csv row objects to import
+   * @memberof CmdbCsvDataSource
+   */
+   constructor(taskAuditRecord, auth_payload, recordType, csvRows) {
+    this.taskAuditRecord = taskAuditRecord;
+    this.auth_payload = auth_payload;
+    this.recordType = recordType;
+    this.csvRows = csvRows;
+
+    // Set initial status
+    this.status = { itemsProcessed: 0, itemTotal: 0, individualRecordStatus: [] };
+  }
+
+  /**
+   * Run the Agri csv importer.
+   *
+   * @returns final status of importer
+   * @memberof CmdbCsvDataSource
+   */
+   async run() {
+    defaultLog.info('run - import agri-cmdb-csv');
+
+    this.status.itemTotal = this.csvRows.length;
+
+    await this.taskAuditRecord.updateTaskRecord({ status: 'Running', itemTotal: this.csvRows.length });
+
+    await this.batchProcessRecords();
+
+    return this.status;
+  }
+
+  /**
+   * Runs processRecord() on each csv row, in batches.
+   *
+   * Batch size configured by env variable `CSV_IMPORT_BATCH_SIZE` if it exists, or 100 by default.
+   *
+   * @memberof CmdbCsvDataSource
+   */
+   async batchProcessRecords() {
+    try {
+      const recordTypeConfig = this.getRecordTypeConfig();
+
+      if (!recordTypeConfig) {
+        throw Error('batchProcessRecords - failed to find matching recordTypeConfig.');
+      }
+
+      let recordInspectionIds = [];
+      let recordOutcomeDescriptions = [];
+
+      // construct the outcome description by appending the regulation sections of rows with the same inspection id
+      for (let i = 0; i < this.csvRows.length; i++) {
+        let inspectionId = this.csvRows[i]['inspection id'];
+        let regulationSection = this.csvRows[i]['regulation section'];
+        let outcomeDescription = 'Compliance issue(s) identified under the following acts or regulations: ' + regulationSection;
+
+        if (recordInspectionIds.includes(inspectionId)) {
+          let index = recordInspectionIds.indexOf(inspectionId);
+          outcomeDescription = recordOutcomeDescriptions[index] + '; ' + regulationSection;
```

Comment:
Duplicate checking for regulationSection would be good to have here. Several of the records have multiple copies of the same regulationSection value.
eg. Company 7 outcomeDescription has 2 FSLR-s.32 values

Reply:
good catch thanks, updated